README.md
# Express Multi Hijack

Simply hijack or intercept your express response via middleware.

## Installation

> npm i express-multi-hijack

## How it works

This library works by overriding `res.end` function, still maintaining the original one once all of the "hijack" functions are executed.

So it is safe to use with Your own custom `res.end` implementation.

## Example


1. Simple Tracing Response Body
```js
const Express = require('express');
const { Hijack } = require('express-multi-hijack');

const app = Express()

app.use(Hijack({
  json: true,
  handler: (body, req, res, next) => {
    console.info('got response body', body) // { "foo": "bar" }
  }
}))

app.get('/bar', (req, res, next) => {
  res.json({
    bar: 'foo',
  })
})

app.get('/foo', (req, res, next) => {
  res.json({
    foo: 'bar',
  })
})

app.listen(8080)
```

Based on the example above, the hijack handler will print `{ "foo": "bar" }` on request to [localhost:8080/foo](http://localhost:8080/foo) and `{ "bar": "foo" }` when requesting to [localhost:8080/bar](http://localhost:8080/bar).

2. Intercept and Override Response Body

```js
const Express = require('express')
const { Hijack } = require('../dist')

const app = Express()

app.use(Hijack({
  json: true,
  handler: (body, _, res) => {
    console.info('body:', body)
    res.json({ message: 'Today is the bad day' })
  }
}))

app.get('/', (_, res) => {
  return res.json({ message: 'Today is a Great Day!' })
})

app.listen(8080)
```

The code above will change the response `message` from `'Today is a Great Day!'` to `'Today is the bad day'`

3. Multiple Hijack, early hijack function is the highest priority

```js
const Express = require('express')
const { Hijack } = require('../dist')

const app = Express()

// hijack on monday
app.use(Hijack({
  json: true,
  handler: (body, _, res) => {
    if (body.today === 'Monday') {
      res.json({
        ...body,
        message: 'What a lazy day'
      })
    }
  }
}))

// hijack on friday
app.use(Hijack({
  json: true,
  handler: (body, req, res) => {
    if (req.path === '/friday') {
      res.json({
        ...body,
        message: 'What a wondeful day'
      })
    }
  }
}))

app.get('/', (_, res) => res.json({ message: 'No Hijack' }))

app.get('/monday', (_, res) => {
  return res.status(200).json({ today: 'Monday', message: 'Today is a Great Day!' })
})

app.get('/friday', (_, res) => {
  return res.status(200).json({ message: 'Today is a Great Day!' })
})

app.listen(8080)

The code above will intercept and change response for [localhost:8080/monday](http://localhost:8080/monday) to `{today: 'Moday', message: 'What a lazy day'}`, and [localhost:8080/friday](http://localhost:8080/friday) to `{message: 'What a wondeful day'}`
```


For more examples please visit [example](https://github.com/slaveofcode/express-multi-hijack/tree/master/example) directory.

## LICENSE
MIT

CHANGELOG.md
# Changelog

## v1.0.1
- Fix Github Action

## v1.0.0
- Hijack with JSON response format
- Hijack with non JSON response format
- Multi attach hijack function
- Added simple README
- Added working examples

src/index.test.ts
import Express from 'express'
import Supertest from 'supertest'
import { Hijack } from './index'

test('Hijack should be able to process response body', async () => {
  const app = Express()

  const fnCall = jest.fn()

  app.use(Hijack({ json: true, handler: (body, _req, _res, _next) => {
    fnCall(body)
  }}))


  app.get('/foo-bar', (_, res) => {
    return res.status(200).json({ message: 'foobar' })
  })

  const api = Supertest(app)

  const resp = await api.get('/foo-bar')

  expect(resp.status).toEqual(200)
  expect(resp.body).toEqual({
    message: 'foobar',
  })
  expect(fnCall).toHaveBeenCalledWith({
    message: 'foobar',
  })
})

test('Hijack should be able to override response body', async () => {
  const app = Express()

  app.use(Hijack({ json: true, handler: (_body, _req, res, _next) => {
    return res.json({ message: 'overrided message' })
  }}))

  app.get('/foo-bar', (_, res) => {
    return res.json({ message: 'foobar' })
  })

  const api = Supertest(app)

  const resp = await api.get('/foo-bar')

  expect(resp.status).toEqual(200)
  expect(resp.body).toEqual({
    message: 'overrided message',
  })
})

test('Hijack should be able to override response body with multiple hijack bottom up', async () => {
  const app = Express()

  app.use(Hijack({ json: true, handler: (_body, _req, res, _next) => {
    return res.json({ message: 'overrided message 2' })
  }}))

  app.use(Hijack({ json: true, handler: (_body, _req, res, _next) => {
    return res.json({ message: 'overrided message 1' })
  }}))

  app.get('/foo-bar', (_, res) => {
    return res.json({ message: 'foobar' })
  })

  const api = Supertest(app)

  const resp = await api.get('/foo-bar')

  expect(resp.status).toEqual(200)
  expect(resp.body).toEqual({
    message: 'overrided message 2',
  })
})

test('Hijack should be able to override for non JSON response with error', async () => {
  const app = Express()

  const errFunCall = jest.fn()

  app.use(Hijack({
    json: true,
    handler: (_body, _req, res, _next) => {
      return res.status(200).send('overrided message').end()
    },
    onJsonParseError: errFunCall,
  }))

  app.get('/foo-bar', (_, res) => {
    return res.json({ message: 'foobar' })
  })

  const api = Supertest(app)

  const resp = await api.get('/foo-bar').responseType('blob')

  expect(resp.status).toEqual(200)
  expect(Buffer.from(resp.body).toString('utf8')).toEqual('overrided message')
  expect(errFunCall).toHaveBeenCalled()
})

test('Hijack should be able to override for non JSON response without error', async () => {
  const app = Express()

  const errFunNoCall = jest.fn()

  app.use(Hijack({
    handler: (_body, _req, res, _next) => {
      return res.status(200).send('overrided message').end()
    },
    onJsonParseError: errFunNoCall,
  }))

  app.get('/foo-bar', (_, res) => {
    return res.json({ message: 'foobar' })
  })

  const api = Supertest(app)

  const resp = await api.get('/foo-bar').responseType('blob')

  expect(resp.status).toEqual(200)
  expect(Buffer.from(resp.body).toString('utf8')).toEqual('overrided message')
  expect(errFunNoCall).not.toHaveBeenCalled()
})


src/index.ts
import { Request, Response, NextFunction } from 'express'
import { nanoid } from 'nanoid'

interface IHijackConfig {
  json?: boolean
  onJsonParseError?: (err: Error) => void | any
  handler: (responseBody: any, req: Request, res: Response, next: NextFunction) => void | any
}

const Hijack = (config: IHijackConfig) => {
  if (config.onJsonParseError && typeof config.onJsonParseError !== 'function') {
    throw new TypeError('"onJsonParseError" must be a function!')
  }

  // @ts-ignore
  const middleware = Middleware(config)
  return middleware
}

const Middleware = (config: IHijackConfig) => (req: Request, res: Response, next: NextFunction) => {
  const defWrite = res.write
  const defEnd = res.end
  const chunks: any[] = []
  const uniqKeyHijack = '__multi_hijack_' + nanoid()

  const nextFn: any = next

  res.write = (...args: any[]) => {
    chunks.push(Buffer.from(args[0]))
    return defWrite.apply(res, [args[0], args[1], args[2]])
  }

  res.end = (...args: any[]) => {
    // if (args[0] && chunks.length > 0) {
    //   chunks = [] // reset
    // }

    if (args[0]) {
      chunks.push(Buffer.from(args[0]))
    }

    const body = Buffer.concat(chunks).toString('utf8')

    let responseBody = body

    if (config.json) {
      try {
        responseBody = JSON.parse(body)
      } catch (err) {
        if (config.onJsonParseError) {
          config.onJsonParseError(err)
        }
      }
    }

    if (!(req as any)[uniqKeyHijack]) {
      (req as any)[uniqKeyHijack] = true
      // if not return any value from function
      // or not calling response inside the function
      // will continue to the original end (defEnd)
      const isOverride = config.handler(responseBody, req, res, next)
      if (!isOverride) {
        defEnd.apply(res, [args[0], args[1], args[2]])
      }
    } else {
      defEnd.apply(res, [args[0], args[1], args[2]])
    }
  }

  return nextFn()
}

export {
  Hijack,
}


example/intercept-multiple.js
const Express = require('express')
const { Hijack } = require('../dist')

const app = Express()

// hijack on monday
app.use(Hijack({
  json: true,
  handler: (body, _, res) => {
    if (body.today === 'Monday') {
      res.json({
        ...body,
        message: 'What a lazy day'
      })
    }
  }
}))

// hijack on friday
app.use(Hijack({
  json: true,
  handler: (body, req, res) => {
    if (req.path === '/friday') {
      res.json({
        ...body,
        message: 'What a wondeful day'
      })
    }
  }
}))

app.get('/', (_, res) => res.json({ message: 'No Hijack' }))

app.get('/monday', (_, res) => {
  return res.status(200).json({ today: 'Monday', message: 'Today is a Great Day!' })
})

app.get('/friday', (_, res) => {
  return res.status(200).json({ message: 'Today is a Great Day!' })
})

console.info('Make a request to http://localhost:8080/monday and http://localhost:8080/friday to see the effect')

app.listen(8080)